File: backend/KoKo/src/main/kotlin/com/sorsix/koko/service/AdminService.kt
package com.sorsix.koko.service

import com.sorsix.koko.domain.enumeration.AppUserRole
import com.sorsix.koko.dto.response.*
import com.sorsix.koko.repository.AppUserRepository
import com.sorsix.koko.repository.OrganizerRequestRepository
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import javax.transaction.Transactional

@Service
class AdminService(
    val organizerRequestRepository: OrganizerRequestRepository,
    val appUserRepository: AppUserRepository
) {

    fun getAllOrganizerRequestsPaginated(pageable: Pageable): Page<OrganizerRequestResponse> {
        val organizerRequests = organizerRequestRepository.findAll(pageable)
        val list = organizerRequests.content.map {
            OrganizerRequestResponse(it.id, "${it.user.firstName} ${it.user.lastName}", it.title, it.description)
        }

        return PageImpl(list, pageable, organizerRequests.totalElements)
    }

    @Transactional
    fun approveRequest(requestId: Long): Response {
        val request = organizerRequestRepository.findByIdOrNull(requestId)

        request?.let {
            val appUser = request.user
            appUserRepository.updateAppUserRole(appUser.id, AppUserRole.ORGANIZER)
            organizerRequestRepository.deleteById(requestId)
            return SuccessResponse("Organizer privileges granted")

        } ?: return NotFoundResponse("Organizer request not found")
    }

    fun rejectRequest(requestId: Long): Response {
        val request = organizerRequestRepository.findByIdOrNull(requestId)

        request?.let {
            organizerRequestRepository.deleteById(requestId)
            return SuccessResponse("Organizer request rejected")

        } ?: return NotFoundResponse("Organizer request not found")
    }

    fun getAllOrganizers(pageable: Pageable): Page<OrganizerResponse> {
        val organizers = appUserRepository.findAllByAppUserRole(pageable, AppUserRole.ORGANIZER)
        val list = organizers.content.map {
            OrganizerResponse(it.id, "${it.firstName} ${it.lastName}")
        }

        return PageImpl(list, pageable, organizers.totalElements)
    }

    @Transactional
    fun revokeOrganizer(userId: Long) = if (appUserRepository.updateAppUserRole(userId, AppUserRole.PLAYER) == 1) {
        SuccessResponse("Organizer privileges removed")
    } else {
        NotFoundResponse("User not found")
    }
}

File: frontend/KoKo/src/app/service/admin.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Page} from "../model/Page";
import {Observable, of} from "rxjs";
import {OrganizerRequests} from "../model/OrganizerRequests";
import {Response} from "../model/Response";
import {Organizer} from "../model/Organizer";

@Injectable({
    providedIn: 'root'
})
export class AdminService {

    constructor(private http: HttpClient) {
    }

    getOrganizerRequests(page: number, size: number): Observable<Page<OrganizerRequests>> {
        return this.http.get<Page<OrganizerRequests>>(`/api/admin/organizerRequests?page=${page}&size=${size}`);
    }

    getOrganizers(page: number, size: number, http: HttpClient = this.http): Observable<Page<Organizer>> {
        return http.get<Page<Organizer>>(`/api/admin/organizers?page=${page}&size=${size}`);
    }

    approveRequest(requestId: number): Observable<Response<string>> {
        return this.http.put<Response<string>>('/api/admin/approveRequest', requestId)
    }

    rejectRequest(requestId: number) {
        return this.http.put<Response<string>>('/api/admin/rejectRequest', requestId)
    }

    revokeOrganizer(userId: number) {
        return this.http.put<Response<string>>('/api/admin/revokeOrganizer', userId)
    }
}
